src/wp-includes/class-wp-scripts.php

```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
-			if ( ! $this->has_only_delayed_dependents( $dependent, $async_only, $checked ) ) {
-				return false;
-			}
+		// Bail early once we know the eligible strategy is blocking.
+		if ( empty( $eligible ) ) {
+			return '';
 		}
 
-		return true;
+		return in_array( 'async', $eligible, true ) ? 'async' : 'defer' ;
 	}
 
 	/**
-	 * Gets the best eligible loading strategy for a script.
+	 * Filter the list of eligible loading strategies for a script.
 	 *
 	 * @since 6.3.0
 	 *
-	 * @param string  $handle The script handle.
-	 * @return string $strategy The best eligible loading strategy.
+	 * @param string              $handle   The script handle.
+	 * @param string[]|null       $eligible Optional. The list of strategies to filter. Default null.
+	 * @param array<string, true> $checked  Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return string[] A list of eligible loading strategies that could be used.
 	 */
-	private function get_eligible_loading_strategy( $handle ) {
+	private function filter_eligible_strategies( $handle, $eligible = null, $checked = array() ) {
+		// If no strategies are being passed, all strategies are eligible.
+		if ( null === $eligible ) {
+			$eligible = $this->delayed_strategies;
+		}
+
+		// If this handle was already checked, return early.
+		if ( isset( $checked[ $handle ] ) ) {
+			return $eligible;
+		}
+
+		// Mark this handle as checked.
+		$checked[ $handle ] = true;
+
+		// If this handle isn't registered, don't filter anything and return.
 		if ( ! isset( $this->registered[ $handle ] ) ) {
-			return '';
+			return $eligible;
+		}
+
+		// If the handle is not enqueued, don't filter anything and return.
+		if ( ! $this->query( $handle, 'enqueued' ) ) {
+			return $eligible;
 		}
 
+		$is_alias = (bool) ! $this->registered[ $handle ]->src;
 		$intended_strategy = $this->get_data( $handle, 'strategy' );
 
-		/*
-		 * Handle known blocking strategy scenarios.
-		 * - An empty strategy is synonymous with blocking.
-		 * - A script alias (where $src is false) must always be blocking since the after inline script cannot be
-		 *   delayed as there is no external script tag and thus no load event at which the inline script can be run.
-		 */
-		if ( empty( $intended_strategy ) ) {
-			return '';
+		// For non-alias handles, an empty intended strategy filters all strategies.
+		if ( ! $is_alias && empty( $intended_strategy ) ) {
+			return array();
 		}
 
-		// Aliases that have before/after inline scripts can never be delayed since there is no load event.
-		if ( ! $this->registered[ $handle ]->src && $this->has_inline_script( $handle ) ) {
-			return '';
+		// Handles with inline scripts attached in the 'after' position cannot be delayed.
+		if ( $this->has_inline_script( $handle, 'after' ) ) {
+			return array();
 		}
 
-		// Handle async strategy scenario.
-		if ( 'async' === $intended_strategy && $this->has_only_delayed_dependents( $handle, true ) ) {
-			return 'async';
```

Comment:
This condition I don't see in the modified logic you implemented. Can you clarify why it was removed? It feels risky to me to make such broad changes so shortly before intended commit and include logic changes like this.

Reply:
I believe this condition is unnecessary after the refactor. Rather than checking for whether all dependents of an async script are also async, we start by assuming async is possible and eliminate that possibility during the process of filtering eligible strategies.

Reply:
Now that I understand how the filtering bit in the `foreach` loop works, that makes sense 👍